azure_data_cosmos_engine/src/query/producer/state.rs

```diff
@@ -64,10 +64,72 @@ impl PartitionState {
             PaginationState::Initial => Some(DataRequest {
                 pkrange_id: self.pkrange.id.clone().into(),
                 continuation: None,
+                query: None,
             }),
             PaginationState::Continuing(token) => Some(DataRequest {
                 pkrange_id: self.pkrange.id.clone().into(),
                 continuation: Some(token.clone()),
+                query: None,
+            }),
+            PaginationState::Done => None,
+        }
+    }
+
+    pub fn update_state(&mut self, continuation: Option<String>) {
+        match continuation {
+            Some(token) => {
+                self.stage = PaginationState::Continuing(token);
+            }
+            None => {
+                self.stage = PaginationState::Done;
+            }
+        }
+    }
+
+    pub fn started(&self) -> bool {
+        !matches!(self.stage, PaginationState::Initial)
+    }
+
+    pub fn done(&self) -> bool {
+        matches!(self.stage, PaginationState::Done)
+    }
+}
+
+#[derive(Debug)]
+pub struct QueryChunkState {
```

Comment:
This is largely the same as `PartitionState`, but with an additional field for the query. How about we merge it with `PartitionState` and just make the `query` an `Option<String>`? It would add a small amount of extra memory to the `PartitionState` type when no query is needed (to store the `None`), but I'm really not worried about that yet (and if we absolutely need to worry about it, there are ways to avoid that).